### PR TITLE
Fixes gpio_callback ambiguity in union pin vs pin_mask

### LINKED ISSUE #12765: drivers: gpio: intel_apl: implementation silently ignores double-edge interrupt config
`gpio_intel_apl.c` properly tests for interrupt trigger and active sense, but does not test for a request to configure double-edge.  It should support this configuration, or return an error indicating the configuration failed.

### LINKED ISSUE #12810: Up Squared serial console character output corrupted
**Describe the bug**
After rebased to latest master branch, observe serial console character output has corrupted.
Tried to revert commit below and found issue no longer there:
f0db0b8815 driver: uart_ns16550.c: necessary changes for arc iot soc

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=up_squared ../
3. make menuconfig
4. select CONFIG_SHELL_BACKEND_SERIAL, CONFIG_LOG
5. make
6. copy newly build zephyr.strip image to usb boot drive
7. boot up Up Squared with the boot drive
8. Observing console printing corrupted

**Expected behavior**
Console printing correctly.

**Impact**
Unable to interact with shell terminal.

**Screenshots or console output**
image

**Environment (please complete the following information):**
 - OS: Ubuntu Linux 14.04
 - Toolchain: Zephyr SDK
 - Commit: 5bd5e4e4406a56c61b4ad0609e3fe238e8ab6864

### PR DESCRIPTION
**Pull request by dcpleung**
- This reverts the patch that original added the "pin" union for GPIO callback. The previous behavior of using pin_mask solely is retained.
- The Apollo Lake GPIO driver has been updated to use only pin_mask for GPIO callback. This requires breaking apart one driver (of 245 pins) into 10 small drivers each controlling at most 32 pins.
- This also fixes #12765 where double-edge interrupt configuration was silently ignore. The datasheet does not specify that double-edge trigger is supported so assign it is not.

### CODE REVIEW

**Comment by pabigot** [comment]
There are at least two problems solved in this PR.  The one with GPIO may take some time to accept; I suggest pulling the UART one out to a separate PR that can get quickly approved.

High level comments on the GPIO one:
* I'm pretty sure `shared-irq` is the wrong name for the new node.  It should be something like `intel,apl-ioapic` though I can only guess at the name based on similar usage in Linux.
* I would have expected to see device-tree nodes for each of the ten GPIO devices.  As it stands I'm not clear how one would reference a specific pin in device tree.

Beyond that I don't have access to this hardware or related experience so I'm not going to try to review the solution.  @galak should probably be involved to confirm the device tree usage is consistent with Linux.

**Comment by dcpleung** [comment]
The shared_irq driver is a virtual interrupt controller. Since the GPIO controller only triggers one interrupt line, the shared_irq is there to distribute it to multiple driver instances. The Apollo Lake hardware still uses IO-APIC for actual hardware interrupts.

As for the number of device nodes, it is still physically one block with one MMIO address. Linux would only have one device node as the pinctrl can access all 245 pins. It is because of the pin_mask in callback that we need 10 driver instances.

File: boards/x86/galileo/galileo.dts [hunk 1]
```diff
@@ -42,3 +42,9 @@
 &spi0 {
 	status = "ok";
 };
+
+#ifdef CONFIG_SHARED_IRQ_0
```
Comment (galak) [inline boards/x86/galileo/galileo.dts]: ifdef's in dts aren't allowed, why do we need this ifdef?

**Review by galak — DISMISSED** [review]
we need a solution that doesn't require an ifdef in the .dts

**Comment by galak** [comment]
> The shared_irq driver is a virtual interrupt controller. Since the GPIO controller only triggers one interrupt line, the shared_irq is there to distribute it to multiple driver instances. The Apollo Lake hardware still uses IO-APIC for actual hardware interrupts.

can you explain what you are trying to do/solve with the shared irq's?

File: boards/x86/galileo/galileo.dts (hunk 1 above)
Comment (dcpleung) [reply boards/x86/galileo/galileo.dts]: Will remove in next rebase.

**Comment by dcpleung** [comment]
@galak The GPIO controller on Apollo Lake has only one interrupt line. And due to the callback pin_mask, one device driver instance in Zephyr can only control at max 32 pins. So multiple driver instances are needed to address all 245 GPIO pins. Since there is only one interrupt line, the shared IRQ driver is used to propagate that one interrupt to all 10 device driver instances.

Another issue is that the controller on APL only triggers interrupt with active low signal. With the previous shared_irq driver design, the trigger condition (CONFIG_SHARED_IRQ_0_LEVEL_LOW) cannot be set in the SoC level. Hence the driver and dts changes.

**Review by tyrantlizard77 — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: boards/x86/galileo/galileo.dts (hunk 1 above)
Comment (dcpleung) [reply boards/x86/galileo/galileo.dts]: Removed the #ifdef in the newest code.

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**